File: lintcode/flatten-binary-tree-to-linked-list.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/flatten-binary-tree-to-linked-list/

"""
Definition of TreeNode:
"""
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
class Solution:
    # @param root: a TreeNode, the root of the binary tree
    # @return: nothing
    def flatten(self, root):
        # write your code here
        if not root:
            return

        self.flatten(root.left)
        self.flatten(root.right)
        prev = None
        curs = root.left
        while curs:
            prev = curs
            curs = curs.right
        if prev:
            prev.right = root.right
            root.right = root.left
            root.left = None

if __name__ == '__main__':
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(5)
    root.left.left = TreeNode(3)
    root.left.right = TreeNode(4)
    root.right.right = TreeNode(6)
    s = Solution()
    s.flatten(root)
    print root.val, root.right.val, root.right.right.val, root.right.right.right.val, root.right.right.right.right.val, root.right.right.right.right.right.val

File: leetcode-2/flatten-binary-tree-to-linked-list.cc
// Question: https://leetcode.com/problems/flatten-binary-tree-to-linked-list/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
        if (!root) {
            return;
        }

        flatten(root->left);
        flatten(root->right);
        TreeNode* curr = root->left;
        TreeNode* prev = NULL;
        while (curr) {
            prev = curr;
            curr = curr->right;
        }
        if (prev) {
            prev->right = root->right;
            root->right = root->left;
            root->left = NULL;
        }
    }

    void flatten(TreeNode* root) {
        auto node = root;
        while (node) {
            if (node->left) {
                auto prev = node->left;
                while (prev->right) {
                    prev = prev->right;
                }
                prev->right = node->right;
                node->right = node->left;
                node->left = NULL;
            }
            // node->right->left = node;
            node = node->right;
        }
    }
};
